<div align="center">

# Modelo de Clasificación del Riesgo de Entrega Tardía en DataCo

### Big Data

**Jonathan Barrantes · Joy D. Chávez Pomacaja · María de Jesús Peña García · Abby Chavarría Bolaños**

Universidad Fidélitas  
San José, Costa Rica

</div>

---

### Autores y contacto

| Integrante                 | Correo institucional        |
| -------------------------- | --------------------------- |
| Jonathan Barrantes         | jbarrantes40180@ufide.ac.cr |
| Joy D. Chávez Pomacaja     | jchavez60736@ufide.ac.cr    |
| María de Jesús Peña García | mpena30226@ufide.ac.cr      |
| Abby Chavarría Bolaños     | achavarria70596@ufide.ac.cr |

---

## Resumen

> Este proyecto busca predecir el riesgo de entregas tardías en DataCo mediante un modelo de clasificación basado en aprendizaje automático. Se utilizará el conjunto de datos **DataCo Smart Supply Chain** y la variable objetivo `Late_delivery_risk`.
>
> Se explorarán y prepararán los datos, se analizarán las características asociadas con los retrasos y se comparará el desempeño de distintos modelos. Las predicciones utilizarán información disponible al registrar el pedido, excluyendo datos posteriores que revelen el resultado de la entrega.
>
> Se pretende generar estimaciones que permitan identificar pedidos con mayor riesgo y apoyar la planificación logística.

**Palabras clave:** aprendizaje automático · cadena de suministro · clasificación · entregas tardías · planificación logística

---

## I. ANTECEDENTES

DataCo Global inició sus operaciones en 2012, dedicada a la comercialización y distribución de productos en el mercado nacional. En sus primeros años, el volumen reducido de pedidos permitía organizar las entregas mediante plazos generales y seguimiento individual.

A partir de 2015, su expansión hacia mercados internacionales incrementó los pedidos, los destinos y las modalidades de envío.

Con este crecimiento, las prácticas iniciales de planificación resultaron insuficientes y comenzaron a presentarse retrasos recurrentes. La empresa atendía los incumplimientos principalmente después de recibir reclamos, lo que dificultaba actuar anticipadamente.

Al mismo tiempo, la competencia y las expectativas de los clientes aumentaron la presión por cumplir los plazos de entrega.

---

## II. DEFINICIÓN DEL PROBLEMA

DataCo presenta limitaciones para anticipar las entregas tardías tras la expansión de sus operaciones. Su planificación continúa utilizando plazos generales que no consideran suficientemente las diferencias entre destinos, modalidades de envío y características de los pedidos.

Además, el seguimiento se concentra en atender incumplimientos después de recibir reclamos, lo que dificulta identificar oportunamente los pedidos con mayor riesgo y priorizar su atención.

Estas limitaciones afectan la planificación de acciones preventivas y pueden ocasionar mayores costos, acumulación de reclamos e insatisfacción de los clientes.

El problema se centra, por tanto, en la dificultad de anticipar posibles retrasos y orientar el seguimiento de los pedidos antes de que se incumplan los plazos.

---

## III. JUSTIFICACIÓN

El proyecto permitiría generar estimaciones del riesgo de entrega tardía para apoyar la planificación logística de DataCo. Esta información facilitaría la identificación de pedidos que requieren mayor atención y la priorización de su seguimiento antes del vencimiento de los plazos.

Asimismo, el análisis de las características de los pedidos, sus destinos y modalidades de envío aportaría criterios para orientar decisiones preventivas.

Estos beneficios podrían contribuir a mejorar la puntualidad de las entregas, controlar los costos asociados con los retrasos y favorecer la satisfacción de los clientes.

---

## IV. OBJETIVO GENERAL

> Desarrollar un modelo de clasificación mediante técnicas de Big Data y aprendizaje automático para predecir el riesgo de entregas tardías en DataCo y apoyar la planificación logística.

---

## V. OBJETIVOS ESPECÍFICOS

1. Analizar las características geográficas, logísticas y de los pedidos presentes en el conjunto de datos, identificando sus relaciones con el riesgo de entrega tardía.

2. Evaluar la calidad de los datos mediante la identificación de valores faltantes, inconsistencias, valores atípicos y distribuciones que puedan afectar el desempeño de los modelos.

3. Preparar los datos mediante técnicas de limpieza, transformación, codificación, escalamiento e imputación, según las necesidades identificadas.

4. Seleccionar y construir variables predictoras relevantes para estimar el riesgo de entrega tardía, considerando únicamente la información disponible al registrar el pedido.

5. Construir distintos modelos de clasificación mediante algoritmos de aprendizaje supervisado para estimar el riesgo de entrega tardía de los pedidos.

6. Comparar el desempeño de los modelos mediante procedimientos de validación y métricas de clasificación, como exactitud, precisión, exhaustividad y F1-score, para determinar su capacidad de identificar pedidos con riesgo de retraso.

7. Ajustar y seleccionar el modelo que presente un equilibrio adecuado entre capacidad predictiva y complejidad, y evaluar su desempeño final sobre un conjunto de prueba reservado para estimar su capacidad de generalización.

---

## VI. DESCRIPCIÓN DEL CONJUNTO DE DATOS

El proyecto utilizará el archivo `DataCoSupplyChainDataset.csv` del conjunto **DataCo Smart Supply Chain**, disponible públicamente en Kaggle [1].

Este contiene información histórica sobre clientes, productos, pedidos y operaciones de envío de DataCo Global.

| Característica                        | Descripción                                                                                                                                                                                            |
| ------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **Estructura**                        | Contiene **180 519 registros y 44 variables** en el archivo utilizado para este proyecto. Las variables abarcan información de clientes, productos, pedidos, ventas y operaciones de envío. |
| **Variable objetivo**                 | `Late_delivery_risk` es una variable binaria que identifica registros etiquetados como entrega tardía mediante `1` y registros no etiquetados como tardíos mediante `0`.                               |
| **Variables categóricas**             | Incluye modalidad de envío, categoría del producto, mercado, región de destino, tipo de transacción y segmento del cliente.                                                                            |
| **Variables numéricas**               | Incluye días programados de envío, cantidad de artículos, precios, descuentos, importes de venta y ganancias.                                                                                          |
| **Variables temporales**              | Contiene fechas de pedido y envío. La fecha de registro del pedido permitirá obtener características como mes y día de la semana.                                                                      |
| **Información posterior a la compra** | Incluye días reales de envío y estados de entrega. Estos campos podrán utilizarse para el análisis histórico y la revisión de las etiquetas, pero se excluirán de las entradas del modelo predictivo.  |

### Variable objetivo

| Valor | Clasificación       |
| :---: | ------------------- |
|  `0`  | Entrega sin retraso |
|  `1`  | Entrega tardía      |

---

## VII. REFERENCIAS

**[1]** S. Tiwari, _DataCo SMART SUPPLY CHAIN FOR BIG DATA ANALYSIS_, Kaggle, conjunto de datos. [En línea]. Disponible en:  
https://www.kaggle.com/datasets/shashwatwork/dataco-smart-supply-chain-for-big-data-analysis  
[Consultado: 22 de septiembre de 2026].


# Análisis exploratorio y preparación inicial de los datos

## 1. Carga y exploración inicial del conjunto de datos


In [1]:
### 1.1 Importación de librerías
import pandas as pd
import numpy as np

In [2]:
### 1.2 Carga del conjunto de datos
from pathlib import Path

rutas_dataset = [
    Path("Dataset/DataCoSupplyChainDataset.csv"),
    Path("DataCoSupplyChainDataset.csv")
]

ruta_dataset = next((ruta for ruta in rutas_dataset if ruta.exists()), None)

if ruta_dataset is None:
    raise FileNotFoundError(
        "No se encontró DataCoSupplyChainDataset.csv. "
        "Colóquelo en la carpeta Dataset/ o junto al notebook."
    )

df = pd.read_csv(
    ruta_dataset,
    encoding="latin1",
    sep=";",
    low_memory=False
)

df.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Name,Customer City,Customer Country,...,Order State,Order Status,Order Zipcode,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,Sporting Goods,Caguas,Puerto Rico,...,Java Occidental,COMPLETE,NaN,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,Sporting Goods,Caguas,Puerto Rico,...,Rajastï¿,PENDING,NaN,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,Sporting Goods,San Jose,EE. UU.,...,Rajastï¿,CLOSED,NaN,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,Sporting Goods,Los Angeles,EE. UU.,...,Queensland,COMPLETE,NaN,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,Sporting Goods,Caguas,Puerto Rico,...,Queensland,PENDING_PAYMENT,NaN,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [3]:
## 1.3 Verificacion de Estructura del CSV
print(f"Registros: {df.shape[0]:,}")
print(f"Variables: {df.shape[1]}")

Registros: 180,519
Variables: 44


In [4]:
### 1.4 Estructura y tipos de datos
resumen_estructura = pd.DataFrame({
    "Variable": df.columns,
    "Tipo de dato": df.dtypes.astype(str).values
})

print(f"Total de variables: {df.shape[1]}")
display(resumen_estructura)

Total de variables: 44


,Variable,Tipo de dato
0,Type,str
1,Days for shipping (real),int64
2,Days for shipment (scheduled),int64
3,Benefit per order,float64
4,Sales per customer,float64
5,Delivery Status,str
6,Late_delivery_risk,int64
7,Category Name,str
8,Customer City,str
9,Customer Country,str


In [5]:
### 1.5 Valores faltantes
resumen_nulos = pd.DataFrame({
    "Valores faltantes": df.isnull().sum(),
    "Porcentaje (%)": (df.isnull().sum() / len(df) * 100).round(2)
})

resumen_nulos = resumen_nulos[resumen_nulos["Valores faltantes"] > 0]
resumen_nulos = resumen_nulos.sort_values(
    by="Valores faltantes",
    ascending=False
)

display(resumen_nulos)

,Valores faltantes,Porcentaje (%)
Product Description,180519,100.00
Order Zipcode,155679,86.24
Customer Lname,8,0.00
Customer Zipcode,3,0.00


Se identificaron valores faltantes en cuatro variables. `Product Description` presenta un **100 % de valores faltantes**, mientras que `Order Zipcode` presenta un **86.24 %**. Por su parte, `Customer Lname` y `Customer Zipcode` presentan únicamente 8 y 3 valores faltantes, respectivamente, por lo que su proporción respecto al total de registros es mínima.

En esta etapa únicamente se documentan los valores faltantes identificados. Su tratamiento se determinará posteriormente durante la preparación de los datos.


In [6]:
### 1.6 Estadísticas descriptivas
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Days for shipping (real),180519.0,3.497654,1.623722,0.000000,2.000000,3.000000,5.000000,6.000000
Days for shipment (scheduled),180519.0,2.931847,1.374449,0.000000,2.000000,4.000000,4.000000,4.000000
Benefit per order,180519.0,21.974989,104.433526,-4274.979980,7.000000,31.520000,64.800003,911.799988
Sales per customer,180519.0,183.107609,120.043670,7.490000,104.379997,163.990005,247.399994,1939.989990
Late_delivery_risk,180519.0,0.548291,0.497664,0.000000,0.000000,1.000000,1.000000,1.000000
Customer Zipcode,180516.0,35921.126914,37542.461122,603.000000,725.000000,19380.000000,78207.000000,99205.000000
Latitude,180519.0,29.719955,9.813646,-33.937553,18.265432,33.144863,39.279617,48.781933
Longitude,180519.0,-84.915675,21.433241,-158.025986,-98.446312,-76.847908,-66.370583,115.263077
Order Item Discount,180519.0,20.664741,21.800901,0.000000,5.400000,14.000000,29.990000,500.000000
Order Item Discount Rate,180519.0,0.101668,0.070415,0.000000,0.040000,0.100000,0.160000,0.250000


Las estadísticas descriptivas muestran diferencias importantes en los rangos y escalas de las variables numéricas. Las variables monetarias presentan una mayor amplitud de valores, mientras que otras variables, como tasas, cantidades y días de envío, se encuentran en rangos considerablemente menores.

También se identifican variables numéricas que no deben interpretarse como continuas, como `Late_delivery_risk`, los códigos postales y `Product Status`. Además, `Product Description` no contiene observaciones válidas y `Product Status` presenta un único valor en todos los registros.

Algunas variables presentan estadísticas descriptivas idénticas, como `Benefit per order` y `Order Profit Per Order`, así como `Sales per customer` y `Order Item Total`. Estas posibles redundancias deberán revisarse posteriormente durante la preparación de los datos.


In [7]:
### 1.7 Identificación de variables continuas
variables_continuas = [
    "Benefit per order",
    "Sales per customer",
    "Latitude",
    "Longitude",
    "Order Item Discount",
    "Order Item Discount Rate",
    "Order Item Product Price",
    "Order Item Profit Ratio",
    "Sales",
    "Order Item Total",
    "Order Profit Per Order",
    "Product Price"
]

print(f"Variables continuas identificadas: {len(variables_continuas)}")
variables_continuas

Variables continuas identificadas: 12


['Benefit per order',
 'Sales per customer',
 'Latitude',
 'Longitude',
 'Order Item Discount',
 'Order Item Discount Rate',
 'Order Item Product Price',
 'Order Item Profit Ratio',
 'Sales',
 'Order Item Total',
 'Order Profit Per Order',
 'Product Price']

## 2. Análisis de distribuciones

**Asignado: Joy D. Chávez Pomacaja**

Generar los histogramas de las variables continuas e identificar cuáles presentan distribuciones simétricas, sesgadas, multimodales, acotadas o uniformes.


## 3. Análisis de escalas y posibles tratamientos


En esta sección se comparan las escalas de las variables numéricas continuas identificadas previamente. El objetivo es reconocer diferencias de magnitud que puedan afectar a ciertos algoritmos de aprendizaje automático, documentar posibles tratamientos para etapas posteriores e identificar una variable adecuada para realizar una división estratificada de los datos.

In [8]:
### 3.1 Comparación de escalas de las variables continuas
variables_escala = [col for col in variables_continuas if col in df.columns]

resumen_escalas = pd.DataFrame({
    "Mínimo": df[variables_escala].min(),
    "Máximo": df[variables_escala].max(),
    "Media": df[variables_escala].mean(),
    "Desviación estándar": df[variables_escala].std(),
    "Q1": df[variables_escala].quantile(0.25),
    "Mediana": df[variables_escala].median(),
    "Q3": df[variables_escala].quantile(0.75),
    "Asimetría": df[variables_escala].skew()
})

resumen_escalas["Rango"] = resumen_escalas["Máximo"] - resumen_escalas["Mínimo"]
resumen_escalas["RIC"] = resumen_escalas["Q3"] - resumen_escalas["Q1"]

columnas_orden = [
    "Mínimo", "Máximo", "Rango", "Media", "Desviación estándar",
    "Q1", "Mediana", "Q3", "RIC", "Asimetría"
]

resumen_escalas = resumen_escalas[columnas_orden].sort_values(
    by="Rango",
    ascending=False
)

display(resumen_escalas.round(4))

,Mínimo,Máximo,Rango,Media,Desviación estándar,Q1,Mediana,Q3,RIC,Asimetría
Benefit per order,-4274.9800,911.8000,5186.7800,21.9750,104.4335,7.0000,31.5200,64.8000,57.8000,-4.7418
Order Profit Per Order,-4274.9800,911.8000,5186.7800,21.9750,104.4335,7.0000,31.5200,64.8000,57.8000,-4.7418
Order Item Product Price,9.9900,1999.9900,1990.0000,141.2325,139.7325,50.0000,59.9900,199.9900,149.9900,3.1910
Sales,9.9900,1999.9900,1990.0000,203.7721,132.2731,119.9800,199.9200,299.9500,179.9700,2.8842
Product Price,9.9900,1999.9900,1990.0000,141.2325,139.7325,50.0000,59.9900,199.9900,149.9900,3.1910
Order Item Total,7.4900,1939.9900,1932.5000,183.1076,120.0437,104.3800,163.9900,247.4000,143.0200,2.8884
Sales per customer,7.4900,1939.9900,1932.5000,183.1076,120.0437,104.3800,163.9900,247.4000,143.0200,2.8884
Order Item Discount,0.0000,500.0000,500.0000,20.6647,21.8009,5.4000,14.0000,29.9900,24.5900,3.0398
Longitude,-158.0260,115.2631,273.2891,-84.9157,21.4332,-98.4463,-76.8479,-66.3706,32.0757,-0.4985
Latitude,-33.9376,48.7819,82.7195,29.7200,9.8136,18.2654,33.1449,39.2796,21.0142,-0.0980


### 3.2 Interpretación de las escalas

La comparación permite observar que las variables no se encuentran en una escala homogénea. Las variables relacionadas con ventas, precios, descuentos y beneficios pueden manejar magnitudes monetarias considerablemente mayores que variables expresadas como tasas o proporciones. `Latitude` y `Longitude`, aunque también son numéricas, representan coordenadas geográficas y deben interpretarse de manera distinta a una medida monetaria.

Estas diferencias de escala no representan un problema para todos los algoritmos. Los modelos basados en árboles suelen ser poco sensibles a la magnitud de las variables, mientras que métodos basados en distancias, gradientes o regularización pueden verse afectados. Por esta razón, cualquier escalado deberá aplicarse posteriormente dentro del proceso de preparación del modelo y ajustarse únicamente con el conjunto de entrenamiento, evitando fuga de información hacia el conjunto de prueba.

En los datos utilizados se observa claramente esta diferencia: `Order Item Discount Rate` varía entre **0 y 0.25**, `Order Item Quantity` entre **1 y 5**, mientras que `Sales` y `Product Price` llegan hasta **1 999.99** y `Benefit per order` se extiende desde **-4 274.98 hasta 911.80**. Esto confirma que las variables manejan rangos muy distintos y justifica documentar tratamientos específicos según el algoritmo que se utilice posteriormente.

In [9]:
### 3.3 Posibles tratamientos futuros
tratamientos_futuros = pd.DataFrame([
    {
        "Grupo de variables": "Ventas, precios y descuentos monetarios",
        "Ejemplos": "Sales, Product Price, Order Item Total, Order Item Discount",
        "Tratamiento posible": "StandardScaler o RobustScaler",
        "Justificación": "Presentan magnitudes mayores y pueden contener valores extremos. RobustScaler puede ser útil si los outliers son importantes."
    },
    {
        "Grupo de variables": "Beneficio y utilidad",
        "Ejemplos": "Benefit per order, Order Profit Per Order",
        "Tratamiento posible": "RobustScaler; transformación con signo si existe fuerte asimetría",
        "Justificación": "En el dataset existen 33 784 valores negativos en estas variables, por lo que no se recomienda aplicar logaritmo directo sin un tratamiento adicional."
    },
    {
        "Grupo de variables": "Tasas y razones",
        "Ejemplos": "Order Item Discount Rate, Order Item Profit Ratio",
        "Tratamiento posible": "Mantener o estandarizar según el algoritmo",
        "Justificación": "Se encuentran en escalas relativamente acotadas; el escalado puede ser innecesario en modelos basados en árboles."
    },
    {
        "Grupo de variables": "Coordenadas geográficas",
        "Ejemplos": "Latitude, Longitude",
        "Tratamiento posible": "Mantener como coordenadas o realizar ingeniería de características geográficas",
        "Justificación": "Representan ubicación y no deben interpretarse de la misma forma que una magnitud monetaria."
    },
    {
        "Grupo de variables": "Códigos postales",
        "Ejemplos": "Customer Zipcode, Order Zipcode",
        "Tratamiento posible": "No escalar como variables continuas",
        "Justificación": "Aunque están almacenados numéricamente, representan códigos de ubicación y no una cantidad medible."
    },
    {
        "Grupo de variables": "Variables sin información útil o constantes",
        "Ejemplos": "Product Description, Product Status",
        "Tratamiento posible": "Evaluar su eliminación",
        "Justificación": "Product Description está vacío en todos los registros y Product Status tiene un único valor (0) en todo el dataset."
    },
    {
        "Grupo de variables": "Variables redundantes",
        "Ejemplos": "Benefit per order / Order Profit Per Order; Sales per customer / Order Item Total",
        "Tratamiento posible": "Revisar y conservar una sola variable de cada par si se confirma redundancia",
        "Justificación": "En el archivo analizado estos pares contienen exactamente los mismos valores, por lo que podrían aportar información duplicada."
    },
    {
        "Grupo de variables": "Variable objetivo",
        "Ejemplos": "Late_delivery_risk",
        "Tratamiento posible": "Mantener como variable binaria 0/1",
        "Justificación": "Es la clase que se desea predecir y no debe normalizarse ni estandarizarse."
    }
])

display(tratamientos_futuros)

,Grupo de variables,Ejemplos,Tratamiento posible,Justificación
0,"Ventas, precios y descuentos monetarios","Sales, Product Price, Order Item Total, Order ...",StandardScaler o RobustScaler,Presentan magnitudes mayores y pueden contener...
1,Beneficio y utilidad,"Benefit per order, Order Profit Per Order",RobustScaler; transformación con signo si exis...,En el dataset existen 33 784 valores negativos...
2,Tasas y razones,"Order Item Discount Rate, Order Item Profit Ratio",Mantener o estandarizar según el algoritmo,Se encuentran en escalas relativamente acotada...
3,Coordenadas geográficas,"Latitude, Longitude",Mantener como coordenadas o realizar ingenierí...,Representan ubicación y no deben interpretarse...
4,Códigos postales,"Customer Zipcode, Order Zipcode",No escalar como variables continuas,"Aunque están almacenados numéricamente, repres..."
5,Variables sin información útil o constantes,"Product Description, Product Status",Evaluar su eliminación,Product Description está vacío en todos los re...
6,Variables redundantes,Benefit per order / Order Profit Per Order; Sa...,Revisar y conservar una sola variable de cada ...,En el archivo analizado estos pares contienen ...
7,Variable objetivo,Late_delivery_risk,Mantener como variable binaria 0/1,Es la clase que se desea predecir y no debe no...


In [10]:
### 3.4 Variables candidatas para estratificación
candidatas_estratificacion = [
    col for col in ["Late_delivery_risk", "Shipping Mode"]
    if col in df.columns
]

resumen_estratificacion = []

for columna in candidatas_estratificacion:
    frecuencias = df[columna].value_counts(dropna=False)
    resumen_estratificacion.append({
        "Variable": columna,
        "Número de grupos": frecuencias.shape[0],
        "Grupo menos frecuente": frecuencias.min(),
        "Proporción mínima (%)": frecuencias.min() / len(df) * 100
    })

resumen_estratificacion = pd.DataFrame(resumen_estratificacion)
display(resumen_estratificacion.round(2))

print("Distribución de Late_delivery_risk:")
display(
    df["Late_delivery_risk"]
    .value_counts(normalize=True)
    .sort_index()
    .mul(100)
    .rename("Porcentaje (%)")
    .to_frame()
    .round(2)
)

,Variable,Número de grupos,Grupo menos frecuente,Proporción mínima (%)
0,Late_delivery_risk,2,81542,45.17
1,Shipping Mode,4,9737,5.39


Distribución de Late_delivery_risk:


,Porcentaje (%)
Late_delivery_risk,
0,45.17
1,54.83


### 3.5 Variable seleccionada para la estratificación

La variable recomendada para la estratificación es `Late_delivery_risk`, debido a que constituye la variable objetivo del problema de clasificación y contiene únicamente dos clases. En el archivo analizado, la clase `0` contiene **81 542 registros (45.17 %)** y la clase `1` contiene **98 977 registros (54.83 %)**. Por lo tanto, existe una distribución suficientemente representativa de ambas clases y es conveniente mantener esas proporciones en entrenamiento y prueba.

`Shipping Mode` también puede considerarse como una variable secundaria de interés porque representa cuatro grupos operativos de envío. Sin embargo, para la división principal se prioriza `Late_delivery_risk`, ya que conservar la distribución de la variable objetivo permite una evaluación más consistente del clasificador.

## 4. Conjuntos de entrenamiento y prueba

Se crearán dos divisiones con una proporción de **80 % para entrenamiento y 20 % para prueba**. La primera será completamente aleatoria y la segunda utilizará estratificación según `Late_delivery_risk`. Ambas emplearán la misma semilla (`random_state=42`) para asegurar reproducibilidad.

In [11]:
### 4.1 Preparación de X e y
from sklearn.model_selection import train_test_split

variable_objetivo = "Late_delivery_risk"

X = df.drop(columns=[variable_objetivo])
y = df[variable_objetivo]

print(f"Registros totales: {len(df):,}")
print(f"Variables predictoras antes de la selección final: {X.shape[1]}")
print(f"Variable objetivo: {variable_objetivo}")

Registros totales: 180,519
Variables predictoras antes de la selección final: 43
Variable objetivo: Late_delivery_risk


In [12]:
### 4.2 División aleatoria
X_train_aleatorio, X_test_aleatorio, y_train_aleatorio, y_test_aleatorio = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42
)

print(f"Entrenamiento aleatorio: {len(X_train_aleatorio):,} registros ({len(X_train_aleatorio) / len(df) * 100:.2f}%)")
print(f"Prueba aleatoria:        {len(X_test_aleatorio):,} registros ({len(X_test_aleatorio) / len(df) * 100:.2f}%)")

Entrenamiento aleatorio: 144,415 registros (80.00%)
Prueba aleatoria:        36,104 registros (20.00%)


In [13]:
### 4.3 División estratificada
X_train_estratificado, X_test_estratificado, y_train_estratificado, y_test_estratificado = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print(f"Entrenamiento estratificado: {len(X_train_estratificado):,} registros ({len(X_train_estratificado) / len(df) * 100:.2f}%)")
print(f"Prueba estratificada:        {len(X_test_estratificado):,} registros ({len(X_test_estratificado) / len(df) * 100:.2f}%)")

Entrenamiento estratificado: 144,415 registros (80.00%)
Prueba estratificada:        36,104 registros (20.00%)


In [14]:
### 4.4 Comparación de proporciones

def obtener_proporciones(serie):
    return (
        serie.value_counts(normalize=True)
        .reindex(sorted(y.dropna().unique()), fill_value=0)
    )

proporciones = pd.DataFrame({
    "Conjunto original": obtener_proporciones(y),
    "Train aleatorio": obtener_proporciones(y_train_aleatorio),
    "Test aleatorio": obtener_proporciones(y_test_aleatorio),
    "Train estratificado": obtener_proporciones(y_train_estratificado),
    "Test estratificado": obtener_proporciones(y_test_estratificado)
}).T * 100

proporciones.columns = [f"Clase {clase} (%)" for clase in proporciones.columns]
display(proporciones.round(4))

,Clase 0 (%),Clase 1 (%)
Conjunto original,45.1709,54.8291
Train aleatorio,45.1719,54.8281
Test aleatorio,45.1667,54.8333
Train estratificado,45.1712,54.8288
Test estratificado,45.1695,54.8305


In [15]:
### 4.5 Error de muestreo respecto a la distribución original
proporcion_original = obtener_proporciones(y)

conjuntos_comparacion = {
    "Train aleatorio": y_train_aleatorio,
    "Test aleatorio": y_test_aleatorio,
    "Train estratificado": y_train_estratificado,
    "Test estratificado": y_test_estratificado
}

resultados_error = []

for nombre, serie in conjuntos_comparacion.items():
    proporcion_split = obtener_proporciones(serie)
    diferencia_pp = (proporcion_split - proporcion_original).abs() * 100

    resultados_error.append({
        "Conjunto": nombre,
        "Error promedio (p.p.)": diferencia_pp.mean(),
        "Error máximo (p.p.)": diferencia_pp.max()
    })

comparacion_error = pd.DataFrame(resultados_error).set_index("Conjunto")
display(comparacion_error.round(6))

,Error promedio (p.p.),Error máximo (p.p.)
Conjunto,,
Train aleatorio,0.001032,0.001032
Test aleatorio,0.004128,0.004128
Train estratificado,0.000340,0.000340
Test estratificado,0.001358,0.001358


In [16]:
### 4.6 Comparación automática de ambas estrategias
error_test_aleatorio = comparacion_error.loc["Test aleatorio", "Error promedio (p.p.)"]
error_test_estratificado = comparacion_error.loc["Test estratificado", "Error promedio (p.p.)"]

print(f"Error promedio del conjunto de prueba aleatorio: {error_test_aleatorio:.6f} puntos porcentuales")
print(f"Error promedio del conjunto de prueba estratificado: {error_test_estratificado:.6f} puntos porcentuales")

if error_test_estratificado <= error_test_aleatorio:
    print("La división estratificada conserva mejor o igual la distribución de la variable objetivo.")
else:
    print("En esta ejecución la división aleatoria presenta un error ligeramente menor, aunque la estratificación garantiza explícitamente la conservación de las clases.")

Error promedio del conjunto de prueba aleatorio: 0.004128 puntos porcentuales
Error promedio del conjunto de prueba estratificado: 0.001358 puntos porcentuales
La división estratificada conserva mejor o igual la distribución de la variable objetivo.


### 4.7 Conclusión de la comparación

La división aleatoria asigna los registros sin imponer restricciones sobre la proporción de cada clase. Debido al tamaño del conjunto de datos (**180 519 registros**), sus proporciones quedan muy cercanas a las del conjunto original; sin embargo, esa similitud depende del muestreo y no está garantizada.

Con `random_state=42` y una división 80/20, el conjunto de prueba aleatorio presenta un error promedio de aproximadamente **0.004128 puntos porcentuales** respecto a la distribución original de `Late_delivery_risk`, mientras que el conjunto de prueba estratificado presenta aproximadamente **0.001358 puntos porcentuales**. Por lo tanto, en este dataset la división estratificada conserva todavía mejor la distribución original.

La estratificación por `Late_delivery_risk` se considera la alternativa más apropiada para este proyecto porque mantiene explícitamente la proporción de entregas tardías y no tardías tanto en entrenamiento como en prueba, haciendo más consistente la evaluación posterior del modelo.